### PR TITLE
create Abonnement model and api, fix boilerplate errors

### PR DESCRIPTION
**Pull request by annashamray**
**Changes:**
1. fix boilerplate errors: 
- change `zds_schema` version
- delete `hijack_tags` from base_site.html
- generate schema
1. create `Abonnement` (subscriber), `Kanaal` (exchange) and `Filters` models
1. create serializers for models and add them to api
1. add test for creating abonnement

### CODE REVIEW

**Comment by sergei-maertens** [comment]
add to whitelist

**Comment by annashamray** [comment]
Changes:
1. Change relations between models:
    - since `Kanaal.name` should be unique `Abonnement` can't be a foreign key to `Kanaal`, one-to-many (foreign key) relations were replaced to many-to-many relations between these models 
    - since `Kanaal.name` should be unique and each `Abonnement` should have it's own `Filters` in it's kanalen, a relation one-to-many (foreign key) was added between `Abonnement` and `Filter`
1. Add api for `Kanaal` which allows GET and POST methods
1. Serializers of `Kanaal` model for `/kanaal` api and for `/abonnement` api were separated
1. Write tests for `/kanaal` api

**Comment by annashamray** [comment]
Changes:
1. add `QueueChannel` class in `api/channels.py` file to connect to RabbitMQ
2. add api for sending notifications - `NotificatieViewSet` and `MessageSerializer`
3. add `FilterGroup` model and change relations between models
4. add tests

File: src/notifications/api/serializers.py [hunk 1]
```diff
@@ -1,19 +1,197 @@
...
+    )
+
+    def validate(self, attrs):
+        validated_attrs = super().validate(attrs)
+        # check if exchange exists
+        try:
+            Kanaal.objects.get(naam=validated_attrs['kanaal'])
+        except ObjectDoesNotExist:
+            raise serializers.ValidationError(
+                {'kanaal': "Kannal with this name doesn't exist"},
```
Comment (joeribekker) [inline src/notifications/api/serializers.py]:
```suggestion
                {'kanaal': _('Kanaal met deze naam bestaat niet.')},
```

File: src/notifications/api/serializers.py [hunk 2]
```diff
@@ -1,19 +1,197 @@
...
+from json import dumps
 
+from django.core.exceptions import ObjectDoesNotExist
+from django.core.serializers.json import DjangoJSONEncoder
+
+import requests
 from rest_framework import serializers
+from rest_framework.response import Response
+
+from notifications.conf.base import CHANNEL
```
Comment (joeribekker) [inline src/notifications/api/serializers.py]: Always import settings via `django.conf.settings`

```suggestion
from django.conf import settings
```

Make sure to change the `CHANNEL` references to `settings.CHANNEL`.

File: src/notifications/api/serializers.py [hunk 3]
```diff
@@ -1,19 +1,197 @@
...
+
+    def validate(self, attrs):
+        validated_attrs = super().validate(attrs)
+        for group_data in validated_attrs['filter_groups']:
+            kanaal_data = group_data['kanaal']
+            try:
+                kanaal = Kanaal.objects.get(naam=kanaal_data['naam'])
+            except ObjectDoesNotExist:
+                raise serializers.ValidationError(
+                    {'naam': "Kannal with this name dosn't exist"},
```
Comment (joeribekker) [inline src/notifications/api/serializers.py]:
```suggestion
                    {'naam': _('Kanaal met deze naam bestaat niet.')},
```

File: src/notifications/api/serializers.py [hunk 4]
```diff
@@ -1,19 +1,197 @@
...
+                'lookup_field': 'uuid',
+            },
+            'auth': {
+                'write_only': True,
+            }
+        }
+
+    def validate(self, attrs):
+        validated_attrs = super().validate(attrs)
+        for group_data in validated_attrs['filter_groups']:
```
Comment (joeribekker) [inline src/notifications/api/serializers.py]: Why is this in the `AbonnementSerializer` and not in the `FilterGroupSerializer`?

No immediate need to change it, but honest question.

File: src/notifications/api/serializers.py [hunk 5]
```diff
@@ -1,19 +1,197 @@
...
+                code='message_kanaal')
+        return validated_attrs
+
+    def _send_to_subs(self, msg):
+        # define subs
+        msg_filters = msg['kenmerken']
+        subs = set()
+        filter_groups = FilterGroup.objects.filter(kanaal__naam=msg['kanaal'])
+        for group in filter_groups:
+            if group.match_pattern(msg_filters):
```
Comment (joeribekker) [inline src/notifications/api/serializers.py]: 👍

File: src/notifications/api/viewsets.py [hunk 6]
```diff
@@ -1,40 +1,78 @@
...
-        'destroy': EXAMPLE_SCOPE,
+        'list': SCOPE_NOTIF_READ_ALL,
+        'retrieve': SCOPE_NOTIF_READ_ALL,
+        'create': SCOPE_NOTIF_CHANGE_ALL,
+        'destroy': SCOPE_NOTIF_CHANGE_ALL,
+        'update': SCOPE_NOTIF_CHANGE_ALL,
+        'partial_update': SCOPE_NOTIF_CHANGE_ALL,
     }
+
+    def create(self, request, *args, **kwargs):
```
Comment (joeribekker) [inline src/notifications/api/viewsets.py]: Why not just define the `MessageSerializer` as serializer on this `NotificatieViewSet`?

File: src/notifications/conf/base.py [hunk 7]
```diff
@@ -336,3 +337,7 @@
             'dsn': RAVEN_CONFIG['dsn']
         },
     })
+
+# RabbitMQ
+BROKER_URL = os.getenv('PUBLISH_BROKER_URL', 'amqp://guest:guest@localhost:5672/%2F')
```
Comment (joeribekker) [inline src/notifications/conf/base.py]: `%2F`?

File: src/notifications/api/tests/test_abonnement.py [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+        abonnement_create_url = get_operation_url('abonnement_create')
+        data = {
+            "callbackUrl": "https://ref.tst.vng.cloud/zrc/api/v1/callbacks",
+            "auth": "Bearer aef34gh",
+            "kanalen": [{
+                "naam": "zaken",
+                "filters": [
+                    {"bron": "082096752011"},
+                    {"zaaktype": "example.com/api/v1/zaaktypen/5aa5c"},
+                    {"vertrouwelijkeidaanduiding": "*"}
```
Comment (joeribekker) [inline src/notifications/api/tests/test_abonnement.py]: Can you search replace `vertrouwelijkeidaanduiding` with `vertrouwelijkheidaanduiding` :)

File: src/notifications/api/serializers.py [hunk 9]
```diff
@@ -1,19 +1,197 @@
...
+        for group_data in validated_data:
+            kanaal_data = group_data.pop('kanaal')
+            filters = group_data.pop('filters')
+
+            kanaal = Kanaal.objects.get(naam=kanaal_data['naam'])
+            filter_group = FilterGroup.objects.create(kanaal=kanaal, abonnement=abonnement)
+            inc = 0
+            for filter in filters:
+                filter.filter_group = filter_group
+                filter.internal_increment = inc
```
Comment (joeribekker) [inline src/notifications/api/serializers.py]: This is not used anymore.

**Comment by annashamray** [comment]
Changes:
1. Correct everything following PR review except `AbonnementSerializer.validate()` and scopes
1. Add small fix to requirements/dev.txt

File: src/notifications/api/viewsets.py (hunk 6 above)
Comment (joeribekker) [reply src/notifications/api/viewsets.py]: Ha, I thought more came out of the box with this. Oh well :)

File: src/notifications/templates/admin/base_site.html [hunk 10]
```diff
@@ -1,5 +1,5 @@
 {% extends "admin/base.html" %}
-{% load i18n static hijack_tags %}
```
[the remark this answers is not in the document]
Comment (joeribekker) [reply src/notifications/templates/admin/base_site.html]: > Sergei indicated this as an ok solution.

I agree.

File: src/notifications/api/serializers.py (hunk 4 above)
Comment (joeribekker) [reply src/notifications/api/serializers.py]: > Anna mentioned due to a bug in vng-api-common, nested serialization goes wrong if it fails.

We choose to be pragmatic and keep it as is, but would be nice if vng-api-common was updated :D

File: src/notifications/conf/base.py (hunk 7 above)
Comment (joeribekker) [reply src/notifications/conf/base.py]: > Anna said this is needed by docs.

ok

**Comment by joeribekker** [comment]
Hi @annashamray Next time, please comment on all review-comments so the reviewer knows why it was, or was not fixed.

**Pull request merged**